### PR TITLE
[ZEPPELIN-2661]Test: personalized mode action

### PR DESCRIPTION
**Pull request by soralee**
### What is this PR for?
Adding to personalized mode action case on Selenium test.
Here is for checking scenario.
#### 1. Simple action test : 
```
1. (admin) login, create a new note, run a paragraph with 'Before' text, turn on personalized mode, logout.
           > checkpoint : result of paragraph is 'Before'.

2. (user1) login, make sure it is on personalized mode and 'Before' in result of paragraph, logout.
           > checkpoint 1 : enabling personalized mode
           > checkpoint 2 : result of paragraph is 'Before'.

3. (admin) login, run after changing paragraph contents to 'After', check result of paragraph, logout.
           > checkpoint : changed from 'Before' to 'After' in paragraph result.

4. (user1) login, check whether result of paragraph is 'Before' or not.
           > checkpoint : result of paragraph is 'Before', 'After' means test fail.
```

#### 2. Graph action test : 
```
1. (admin) login, create a new note, run a paragraph with data of spark tutorial, change active graph to 'Bar Chart', turn on personalized mode, logout.
           > checkpoint 1: result string of paragraph contains import 'org.apache.commons.io.IOUtils'.
           > checkpoint 2 : check active graph is 'Bar Chart'

2. (user1) make sure it is on personalized mode and graph mode is on 'Bar Chart', try to change active graph to 'Table' and then check result
           > checkpoint 1 : enabling personalized mode
           > checkpoint 2 : final result of active graph is 'Bar Chart', 'Table' means test fail.
```

#### 3. Dynamic form action test : 
```
1. (admin) login, create a new note, run a paragraph with data of spark tutorial, logout.
           > checkpoint 1: result string of paragraph contains import 'org.apache.commons.io.IOUtils'.
           > checkpoint 2 : check a dynamic form value is '30'

2. (user1) make sure it is on personalized mode and  dynamic form value is '30', try to change dynamic form value to '10' and then check result
           > checkpoint 1 : enabling personalized mode
           > checkpoint 2 : final result of dynamic form value is '30', '10' means test fail.
```

### What type of PR is it?
[ Test ]

### What is the Jira issue?
* [ZEPPELIN-2661](https://issues.apache.org/jira/browse/ZEPPELIN-2661)

### How should this be tested?
- build zeppelin with `mvn clean package -DskipTests'

 -  1. execute for Simple text test : `TEST_SELENIUM="true" mvn test -pl 'zeppelin-server' --am -DfailIfNoTests=false -Dtest=PersonalizeActionsIT#testSimpleAction`

 - 2. execute for Graph test : `TEST_SELENIUM="true" mvn test -pl 'zeppelin-server' --am -DfailIfNoTests=false -Dtest=PersonalizeActionsIT#testGraphAction`

 - 3. execute for Graph test : `TEST_SELENIUM="true" mvn test -pl 'zeppelin-server' --am -DfailIfNoTests=false -Dtest=PersonalizeActionsIT#testDynamicFormAction`

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? No

### CODE REVIEW

**Comment by soralee** [comment]
@jongyoul @1ambda please help to review :slightly_smiling_face:

**Comment by 1ambda** [comment]
Could you make CI green?

**Comment by soralee** [comment]
reopen for CI

File: zeppelin-server/src/test/java/org/apache/zeppelin/integration/PersonalizeActionsIT.java [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+  }
+
+  private void logoutUser(String userName) throws URISyntaxException {
+    ZeppelinITUtils.sleep(500, false);
+    driver.findElement(By.xpath("//div[contains(@class, 'navbar-collapse')]//li[contains(.,'" +
+        userName + "')]")).click();
+    ZeppelinITUtils.sleep(500, false);
+    driver.findElement(By.xpath("//div[contains(@class, 'navbar-collapse')]//li[contains(.,'" +
+        userName + "')]//a[@ng-click='navbar.logout()']")).click();
+    ZeppelinITUtils.sleep(2000, false);
```
Comment (1ambda) [inline zeppelin-server/src/test/java/org/apache/zeppelin/integration/PersonalizeActionsIT.java]: Can we avoid sleep? It might be a flaky test.

Comment (soralee) [inline zeppelin-server/src/test/java/org/apache/zeppelin/integration/PersonalizeActionsIT.java]: Thanks for review!
Okay, Let me fix this.

**Comment by soralee** [comment]
@1ambda, changed some code so that using minimum `sleep` method. I tried to change all sleep, but I checked If changing page, it needed a little `sleep`.

**Comment by soralee** [comment]
reopen for CI

File: zeppelin-server/src/test/java/org/apache/zeppelin/integration/PersonalizeActionsIT.java [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.List;
+
+import static org.junit.Assert.assertTrue;
+
+
+/**
+ * Created for org.apache.zeppelin.integration on 13/06/16.
```
Comment (1ambda) [inline zeppelin-server/src/test/java/org/apache/zeppelin/integration/PersonalizeActionsIT.java]: I think we don't need this comment.

File: zeppelin-server/src/test/java/org/apache/zeppelin/integration/PersonalizeActionsIT.java [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+      "shiro.loginUrl = /api/login\n" +
+      "[roles]\n" +
+      "admin = *\n" +
+      "user = *\n" +
+      "[urls]\n" +
+      "/api/version = anon\n" +
+      "/** = authc";
+
+  static String originalShiro = "";
+
```
Comment (1ambda) [inline zeppelin-server/src/test/java/org/apache/zeppelin/integration/PersonalizeActionsIT.java]: unnecessary newline

**Comment by 1ambda** [comment]
Generally looks good. This PR is for testing the basic personalized mode behavior what we should have to verify before every release.

AFAIK, outputs are personalized for

- graph (e.g changing graph type)
- dynamic forms

Can we make the similar tests for these cases?

File: zeppelin-server/src/test/java/org/apache/zeppelin/integration/PersonalizeActionsIT.java (hunk 3 above)
Comment (soralee) [inline zeppelin-server/src/test/java/org/apache/zeppelin/integration/PersonalizeActionsIT.java]: you'er right! Let me update again.

File: zeppelin-server/src/test/java/org/apache/zeppelin/integration/PersonalizeActionsIT.java (hunk 2 above)
Comment (soralee) [inline zeppelin-server/src/test/java/org/apache/zeppelin/integration/PersonalizeActionsIT.java]: oops, let me update again.

**Comment by soralee** [comment]
@1ambda Thanks for review and Sounds good! 
Let me make two scenarios which is based above code.

**Comment by soralee** [comment]
Hi, there.
I just added two test cases as the @1ambda comment and have written PR description as well.

**Comment by jongyoul** [comment]
Simple action test LGTM.

**Comment by jongyoul** [comment]
We'd better divide this test into two or three issues as it's not passed by some minor bugs

**Comment by zjffdu** [comment]
Could we run it many times to make sure this is not flaky test ? Sadly, I see flaky test in selenium test very often.

**Comment by soralee** [comment]
@jongyoul sounds good! 
I'll created JIRA ticket to improve personalized mode and test, after merge.

@zjffdu in my local, it works very well. and even my CI is failing other selenium. one thing, I can test in Travis many times. so let me test more on Travis.

**Comment by soralee** [comment]
@zjffdu, yeah, I'm also facing tavis failure so many times as well. This is not a problem that can be fixed with this PR. It's time to that we should take care of overall selenium test. 

here is one thought in my head to quickly resolve issue, how about spliting the 3rd job in Travis? I guess many people keep failing the 2nd and 3rd jobs, because it tooks long time than other job.
And then, we can improve overall selenium for avoid flaky test.

what do you think of?

**Comment by zjffdu** [comment]
Yeah, make sense. Let's fix the flaky test step by step.

**Comment by soralee** [comment]
@1ambda, I updated regarding login, logout code because I think that it would be better to reuse in AuthenticationIT code and to resolve passing CI test. 

Regarding CI failure seems like related with`sleep` code.
I just want to let you know about `sleep` code, because you don't want to use it and it's flaky code. I agree that but, Travis often gives error messages if there are not `sleep`. We should improve this part but, it would be better to fixed when we improve overall selenium test code or split.

what do you think about this?

**Comment by soralee** [comment]
reopen CI becoming green

**Comment by soralee** [comment]
please review again @jongyoul :slightly_smiling_face:

**Comment by jongyoul** [comment]
LGTM. Merge if there's no more discussion

**Pull request closed without merging**